src/App.php

```diff
@@ -987,47 +1074,90 @@ function () {
 
     /* RESPONSES */
 
+    /** @var string[] */
+    private static $_sentHeaders = [];
+
+    /** @var string[] */
+    private static $_x = [];
+
     /**
-     * Output Response to the client with custom headers.
+     * Output Response to the client.
      *
      * This can be overridden for future PSR-7 implementation
-     *
-     * @TODO SSE is a "Header in Header" case, it works, but must be checked
      */
-    protected function outputResponse(array $headers, $content)
+    protected function outputResponse(string $data, array $headers): void
     {
-        // if header already sent don't send header
-        // @TODO check this, because in theory multiple header sent
-        // can be a symptom of wrong usage
-        if (!headers_sent()) {
-            foreach ($headers as $header => $replace) {
-                header($header, $replace);
-            }
+        $this->response_headers = $this->normalizeHeaders($this->response_headers);
+        $headersAll = array_merge($this->response_headers, $this->normalizeHeaders($headers));
+        $headersNew = array_diff_assoc($headersAll, self::$_sentHeaders);
+
+        if (count($headersNew) > 0) {
+            ob_start();
+            debug_print_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+            $trace = ob_get_contents();
+            ob_clean();
+            ob_end_flush();
+            self::$_x[] = [$headersNew, $trace, headers_list(), http_response_code()];
         }
 
-        echo $content;
+        if (count($headersNew) > 0 && headers_sent()) {
+            echo "\n" . '!! ATK4 UI ERROR: Headers already sent, more headers can not be set at this stage. !!' . "\n";
+
+            // debug
+            echo '--> Mvorisek:' . "\n";
```

Comment:
should this be cleaned up?

Reply:
@romaninsh Yes, it should, see my comment:

> Based on PR #1065. Rebase once that PR is merged.

So please review the PR #1065 first :)